lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -0,0 +1,950 @@
+// to a type instead of a different type of declaration.
+static Type lookupFoundationType(ASTContext &C, Identifier name) {
+  return lookupFoundationTypeDecl(C, name)->getDeclaredInterfaceType();
+}
+
+// Validates that all the variables declared in the given list of declarations
+// conform to the Codable protocol.
+//
+// Produces a diagnostic on the given typechecker for every var which does not
+// conform. Calls a success callback for every var which does conform.
+template <typename ValidVarCallback>
+static bool
+validateVarsConformToCodable(TypeChecker &tc, DeclContext *context,
+                             NominalTypeDecl::StoredPropertyRange &vars,
+                             ValidVarCallback &callback) {
+  auto codableProto = tc.Context.getProtocol(KnownProtocolKind::Codable);
+  bool allConform = true;
+  for (auto varDecl : vars) {
+    // If the decl doesn't yet have a type, we may be seeing it before the type
+    // checker has gotten around to evaluating its type. For example:
+    //
+    // func foo() {
+    //   let b = Bar(from: decoder) // <- evaluates Bar conformance to Codable,
+    //                              //    forcing derivation
+    // }
+    //
+    // struct Bar : Codable {
+    //   var x: Int // <- we get to valuate x's var decl here, but its type
+    //              //    hasn't yet been evaluated
+    // }
+    //
+    // Validate the decl eagerly.
+    if (!varDecl->hasType())
+      tc.validateDecl(varDecl);
+
+    // If the var decl didn't validate, it may still not have a type; confirm it
+    // has a type before ensuring the type conforms to Codable.
+    if (varDecl->hasType() &&
+        tc.conformsToProtocol(varDecl->getType(), codableProto, context,
+                              ConformanceCheckFlags::Used)) {
+      callback(varDecl);
+    } else {
+      // TODO: We should produce a diagnostic note here explaining that we found
+      //       a var not conforming to Codable.
+      allConform = false;
+    }
+  }
+
+  return allConform;
+}
+
+// Validates the given CodingKeys enum decl by ensuring its cases are a 1-to-1
+// match with the stored vars of the given type.
+static bool
+validateCodingKeysEnum(TypeChecker &tc, EnumDecl *codingKeysDecl,
+                       NominalTypeDecl *type) {
+  // Look through all var decls in the given type.
+  // * Filter out lazy/computed vars (currently already done by
+  //   getStoredProperties).
+  // * Filter out ones which are present in the given decl (by name).
+  //
+  // If any of the entries in the CodingKeys decl are not present in the type
+  // by name, then this decl doesn't match.
+  // If there are any vars left in the type, then this decl doesn't match.
+  //
+  // NOTE: If we change the behavior to ignore vars with default values, then we
+  //       can further filter out the type names to remove those which
+  //       correspond to vars with default values.
+  llvm::DenseSet<Identifier> names;
```

Comment:
Yea... it surprises a lot of people. The nice thing about SmallDenseSet is that IIRC if you have a pointer, it "automagically" reduces down to a SmallPtrSet which is even more efficient.